subprojects/core/src/main/java/org/gradle/api/internal/DependencyInjectingInstantiator.java

```diff
@@ -101,26 +100,30 @@ public CachedConstructor transform(Class<?> aClass) {
         Object[] resolvedParameters = new Object[parameterTypes.length];
         int pos = 0;
         for (int i = 0; i < resolvedParameters.length; i++) {
+            Object currentParameter = null;
             Class<?> targetType = parameterTypes[i];
             if (targetType.isPrimitive()) {
                 targetType = JavaReflectionUtil.getWrapperTypeForPrimitiveType(targetType);
             }
+            Type serviceType = constructor.getGenericParameterTypes()[i];
             if (pos < parameters.length && targetType.isInstance(parameters[pos])) {
-                resolvedParameters[i] = parameters[pos];
+                currentParameter = parameters[pos];
                 pos++;
             } else {
-                Type serviceType = constructor.getGenericParameterTypes()[i];
-                try {
-                    resolvedParameters[i] = services.get(serviceType);
-                } catch (UnknownServiceException use) {
-                    StringBuilder builder = new StringBuilder(String.format("Unable to determine argument #%s: no service of type %s", i, serviceType));
-                    if (pos < parameters.length) {
-                        builder.append(String.format(", or value %s not assignable to type %s", parameters[pos], parameterTypes[i]));
-                    } else {
-                        builder.append(String.format(", or missing parameter value of type %s", parameterTypes[i]));
-                    }
-                    throw new IllegalArgumentException(builder.toString());
+                currentParameter = services.find(serviceType);
+            }
+
+            if (currentParameter != null) {
+                resolvedParameters[i] = currentParameter;
+            } else {
+                StringBuilder builder = new StringBuilder(String.format("Unable to determine argument #%s:", i));
```

Comment:
Shouldn't this be `#%d` instead of `#%s`? And shouldn't it be `i + 1`, so that we count arguments from `1` not `0`?

Reply:
I think we should say what argument we are talking about here. Something like `Unable to determine constructor argument #3 of 'com.example.MyType'`.

Reply:
i'm always leery of `%d` because you won't get any compiler error if the parameter type changes & will then fail at runtime, so i've typically found `%s` to be the safest.

Reply:
updated to use `i + 1`